src/Disciplina/Educator/SizedMerkleTree.hs

```diff
@@ -0,0 +1,237 @@
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE DeriveFoldable #-}
+{-# LANGUAGE DeriveAnyClass #-}
+
+-- | Sized Merkle tree implementation.
+module Disciplina.Educator.SizedMerkleTree
+       ( SizedMerkleRoot(..)
+       , SizedMerkleTree (..)
+       , getSizedMerkleRoot
+       , mkSizedMerkleTree
+
+       , SizedMerkleNode (..)
+       , mkBranch
+       , mkLeaf
+       ) where
+
+import Universum hiding (foldMap, toList)
+
+import Disciplina.Crypto (Hash, HasHash, hash, unsafeHash)
+import Disciplina.Crypto.Hash.Class (HashFunc, AbstractHash (..))
+
+import Codec.Serialise (Serialise(..), serialise)
+import Data.Bits (Bits (..))
+import Data.ByteArray (ByteArrayAccess, convert)
+import Data.ByteString.Builder (Builder, byteString, lazyByteString)
+import Data.Foldable (Foldable (..))
+import Prelude (show)
+
+import qualified Data.ByteString.Builder.Extra as Builder
+import qualified Data.ByteString.Lazy as LBS
+
+-- | Data type for root of sized merkle tree.
+data SizedMerkleRoot a = SizedMerkleRoot
+    { smrHash :: Hash LBS.ByteString  -- ^ returns root 'Hash' of Merkle Tree
+    , smrSize :: Word8 -- ^ size of root node,
+                       -- size is defined as number of leafs in this subtree
+    } deriving (Show, Eq, Ord, Generic, ByteArrayAccess , Serialise, Typeable)
+
+
+data SizedMerkleTree a
+  = SizedMerkleEmpty
+  | SizedMerkleTree !(SizedMerkleNode a)
+  deriving (Eq, Generic)
+
+
+instance Foldable SizedMerkleTree where
+    foldMap _ SizedMerkleEmpty      = mempty
+    foldMap f (SizedMerkleTree n) = foldMap f n
+
+    null SizedMerkleEmpty = True
+    null _                = False
+
+    length SizedMerkleEmpty      = 0
+    length (SizedMerkleTree n) = fromInteger (toInteger (smrSize (mRoot n)))
+
+instance Show a => Show (SizedMerkleTree a) where
+  show SizedMerkleEmpty = "Empty Sized Merkle tree "
+  show (SizedMerkleTree n) = "Sized Merkle tree:  " <> Prelude.show n
+  --show tree = "Merkle tree: "  <> Prelude.show (toList tree)
+
+type LeafIndex = Int
+
+data SizedMerkleNode a
```

Comment:
Instead, you could have `Node = (Info, Node, Node)` and `Leaf a = (a, hash)`, which would make it possible to inline `SizedMerkleRoot` into `Node` constructor directly. `smrSize` is always 1, so it's a redundant info.

Reply:
Is `size` the only example? Then implement a separate function for it :)
I believe that's an easy tradeoff to figure out.

Reply:
@volhovm I think we should have separate `SizedMerkleRoot` datatype still, because these roots are meant to be published and used by clients without any branches.